File: DuAnTauHoaCom/backend/src/services/BookingService.js
const { sequelize } = require('../config/database')
const { Op } = require('sequelize')

// Format JS Date → SQL Server DATETIME-compatible literal (no timezone offset)
const fmtDt = (d) => sequelize.literal(`'${new Date(d).toISOString().replace('T', ' ').slice(0, 23)}'`)
const {
  DonDatVe, Ve, HanhKhach, TamGiuGhe, KhuyenMai,
  ThanhToan, HoaDon, ChuyenTau, LichChay, GaTau, Tau,
} = require('../models')
const BookingRepo = require('../repositories/BookingRepository')
const { genBookingCode, genOrderCode, genTransactionCode, genInvoiceNumber } = require('../utils/helpers')
const { getLichChayId, getThuTu, createGheChangForVe, ensureGheChuyen } = BookingRepo

// Giữ ghế tạm khi user bắt đầu điền form (trước khi đặt vé chính thức)
const holdSeatsForCheckout = async ({ trips }) => {
  const sessionId = require('crypto').randomBytes(8).toString('hex').toUpperCase()
  const hetHan    = new Date(Date.now() + 15 * 60 * 1000)

  for (const trip of trips) {
    const toaMap = {}
    trip.passengerSeats.forEach(ps => {
      if (!toaMap[ps.soToaThuTu]) toaMap[ps.soToaThuTu] = []
      toaMap[ps.soToaThuTu].push(ps.seatNumber)
    })
    for (const [soToa, soGheList] of Object.entries(toaMap)) {
      const ok = await BookingRepo.checkSeatsAvailable(
        trip.idChuyen, parseInt(soToa), soGheList,
        trip.idGaLen, trip.idGaXuong
      )
      if (!ok) throw { status: 409, message: `Ghế đã được đặt trong chuyến #${trip.idChuyen}. Vui lòng chọn lại.` }
    }
    for (const [soToa, soGheList] of Object.entries(toaMap)) {
      await BookingRepo.holdSeats(
        trip.idChuyen, parseInt(soToa), soGheList, sessionId, null,
        trip.idGaLen || null, trip.idGaXuong || null
      )
    }
  }

  return { sessionId, hetHan }
}

// Tạo đơn đặt vé (gồm giữ ghế + tạo HanhKhach + Ve)
// trips[i].sessionId (tuỳ chọn): nếu có, link TamGiuGhe hiện tại thay vì tạo mới
const createBooking = async ({ trips, passengers, contactInfo, idTaiKhoan = null, maKhuyenMai = null }) => {
  return sequelize.transaction(async (t) => {
    // 1. Validate khuyến mãi nếu có
    let khuyenMai = null
    let tienGiam = 0
    if (maKhuyenMai) {
      khuyenMai = await KhuyenMai.findOne({
        where: { ma_khuyen_mai: maKhuyenMai.toUpperCase() },
        transaction: t,
      })
      if (!khuyenMai) throw { status: 400, message: 'Mã khuyến mãi không hợp lệ' }
      const today = new Date().toISOString().slice(0, 10)
      if (today < khuyenMai.ngay_bat_dau || today > khuyenMai.ngay_het_han)
        throw { status: 400, message: 'Mã khuyến mãi đã hết hạn' }
      if (khuyenMai.so_luong && khuyenMai.da_dung >= khuyenMai.so_luong)
        throw { status: 400, message: 'Mã khuyến mãi đã hết lượt sử dụng' }
    }

    // 2. Tính tổng tiền
    const tongTienVe = trips.reduce((s, trip) =>
      s + trip.passengerSeats.reduce((ps, p) => ps + p.seatPrice, 0), 0)
    const phiDichVu = trips.length > 1 ? 40000 : 20000
    const tongTruocGiam = tongTienVe + phiDichVu

    if (khuyenMai && tongTruocGiam >= parseFloat(khuyenMai.gia_tri_don_toi_thieu)) {
      if (khuyenMai.loai_giam === 'phan_tram') {
        tienGiam = Math.floor(tongTruocGiam * parseFloat(khuyenMai.gia_tri) / 100)
        if (khuyenMai.giam_toi_da) tienGiam = Math.min(tienGiam, parseFloat(khuyenMai.giam_toi_da))
      } else {
        tienGiam = Math.min(parseFloat(khuyenMai.gia_tri), tongTruocGiam)
      }
    }
    const tongThanhToan = tongTruocGiam - tienGiam

    // 3. Kiểm tra ghế còn trống (segment-aware)
    for (const trip of trips) {
      const tripSessionId = trip.sessionId || null
      const toaMap = {}
      trip.passengerSeats.forEach(ps => {
        const soToa = ps.soToaThuTu ?? trip.soToaThuTu
        if (!toaMap[soToa]) toaMap[soToa] = []
        toaMap[soToa].push(ps.seatNumber)
      })
      for (const [soToa, soGheList] of Object.entries(toaMap)) {
        const available = await BookingRepo.checkSeatsAvailable(
          trip.idChuyen, parseInt(soToa), soGheList,
          trip.idGaLen || null, trip.idGaXuong || null, tripSessionId
        )
        if (!available) throw { status: 409, message: `Ghế đã được đặt trong chuyến #${trip.idChuyen}. Vui lòng chọn lại.` }
      }
    }

    // 4. Sinh mã đơn
    let maDatCho, maDon
    do { maDatCho = genBookingCode() } while (await DonDatVe.findOne({ where: { ma_dat_cho: maDatCho }, transaction: t }))
    maDon = genOrderCode()

    const hetHan = new Date(Date.now() + 15 * 60 * 1000)

    // 5. Tạo DonDatVe
    const don = await DonDatVe.create({
      ma_don:          maDon,
      ma_dat_cho:      maDatCho,
      id_tai_khoan:    idTaiKhoan,
      ho_ten_lien_lac: contactInfo.hoTen || passengers[0]?.hoTen || '',
      email_dat_cho:   contactInfo.email,
      sdt_dat_cho:     contactInfo.phone.replace(/\s/g, ''),
      cccd:            passengers[0]?.cccd || '000000000',
      loai_ve:         trips.length > 1 ? 'khu_hoi' : 'mot_chieu',
      tong_tien:       tongTruocGiam,
      tien_giam:       tienGiam,
      tien_thanh_toan: tongThanhToan,
      id_khuyen_mai:   khuyenMai?.id_khuyen_mai || null,
      trang_thai:      'cho_thanh_toan',
      thoi_gian_dat:   fmtDt(new Date()),
      thoi_gian_het_han: fmtDt(hetHan),
    }, { transaction: t })

    // 6. Tạo HanhKhach + Ve cho mỗi chuyến
    const veList = []
    let passengerIdx = 0   // Chỉ số hành khách tuần tự qua tất cả các chuyến
    for (const trip of trips) {
      for (let i = 0; i < trip.passengerSeats.length; i++) {
        const ps = trip.passengerSeats[i]
        const p  = passengers[passengerIdx++] ?? passengers[passengers.length - 1]

        // Tìm hoặc tạo HanhKhach
        // Xác định loại hành khách → lưu đúng vào DB
        const loaiHK = p.isChild   ? 'tre_em'
                    : p.isElderly  ? 'nguoi_cao_tuoi'
                    : p.isStudent  ? 'sinh_vien'
                    : 'nguoi_lon'

        let hk = await HanhKhach.findOne({
          where: { ho_ten: p.hoTen, ngay_sinh: p.ngaySinh },
          transaction: t,
        })
        if (!hk) {
          hk = await HanhKhach.create({
            id_tai_khoan:    idTaiKhoan,
            ho_ten:          p.hoTen,
            ngay_sinh:       p.ngaySinh,
            cccd:            p.cccd || null,
            loai_hanh_khach: loaiHK,
            so_dien_thoai:   contactInfo.phone.replace(/\s/g, ''),
            la_chinh:        passengerIdx === 1,
          }, { transaction: t })
        }

        const ve = await Ve.create({
          id_don_dat_ve:    don.id_don_dat_ve,
          id_hanh_khach:    hk.id_hanh_khach,
          id_chuyen:        trip.idChuyen,
          so_toa_thu_tu:    ps.soToaThuTu ?? trip.soToaThuTu,
          so_ghe_trong_toa: ps.seatNumber,
          id_ga_len:        trip.idGaLen,
          id_ga_xuong:      trip.idGaXuong,
          loai_hanh_khach:  loaiHK,
          gia_ve:           ps.seatPrice,
          trang_thai:       'cho_xac_nhan',
          ngay_xuat_ve:     fmtDt(new Date()),
        }, { transaction: t })

        // Tạo / cập nhật GheChang cho ve vừa tạo
        if (trip.idGaLen && trip.idGaXuong) {
          const idLC = await getLichChayId(trip.idChuyen)
          const thuTuTu  = idLC ? await getThuTu(idLC, trip.idGaLen)  : null
          const thuTuDen = idLC ? await getThuTu(idLC, trip.idGaXuong) : null
          const tripSessionId = trip.sessionId || null
          await createGheChangForVe(ve, thuTuTu, thuTuDen, tripSessionId, t)
        }

        veList.push(ve)
      }

      // Giữ ghế tạm thời (nhóm theo toa)
      const holdToaMap = {}
      trip.passengerSeats.forEach(ps => {
        const soToa = ps.soToaThuTu ?? trip.soToaThuTu
        if (!holdToaMap[soToa]) holdToaMap[soToa] = []
        holdToaMap[soToa].push(ps.seatNumber)
      })
      const tripSessionId = trip.sessionId || null
      for (const [soToa, soGheList] of Object.entries(holdToaMap)) {
        if (tripSessionId) {
          // Ghế đã được giữ từ bước chọn ghế — link vào DonDatVe mới tạo
          await TamGiuGhe.update(
            {
              id_don_dat_ve:     don.id_don_dat_ve,
              id_ga_len:         trip.idGaLen || null,
              id_ga_xuong:       trip.idGaXuong || null,
              thoi_gian_het_han: fmtDt(new Date(Date.now() + 15 * 60 * 1000)),
            },
            {
              where: {
                id_chuyen:        trip.idChuyen,
                so_toa_thu_tu:    parseInt(soToa),
                so_ghe_trong_toa: { [Op.in]: soGheList },
                session_id:       tripSessionId,
                trang_thai:       'dang_giu',
              },
              transaction: t,
            }
          )
        } else {
          await BookingRepo.holdSeats(
            trip.idChuyen, parseInt(soToa), soGheList, maDatCho, don.id_don_dat_ve,
            trip.idGaLen || null, trip.idGaXuong || null
          )
        }
      }
    }

    // 7. Cập nhật số lần dùng khuyến mãi
    if (khuyenMai) {
      await KhuyenMai.increment('da_dung', { by: 1, where: { id_khuyen_mai: khuyenMai.id_khuyen_mai }, transaction: t })
    }

    return { don, maDon, maDatCho, tongThanhToan, tienGiam, veList }
  })
}

// Tra cứu đơn đặt vé (theo mã đặt chỗ + email + phone)
const lookupBooking = async (maDatCho, email = '', phone = '') => {
  let don = await BookingRepo.findByMaDatCho(maDatCho)
  if (!don) return null

  // Xác minh cả email VÀ sdt (cả 3 trường phải khớp)
  const emailMatch = don.email_dat_cho.toLowerCase() === email.trim().toLowerCase()
  const phoneClean = don.sdt_dat_cho.replace(/\D/g, '')
  const inputPhone  = phone.trim().replace(/\D/g, '')
  const phoneMatch  = phoneClean === inputPhone || phoneClean.slice(-9) === inputPhone.slice(-9)

  if (!emailMatch || !phoneMatch) return null

  // Nếu đơn chưa thanh toán và đã hết hạn → giải phóng ngay trong DB
  if (don.trang_thai === 'cho_thanh_toan') {
    const hetHanMs = new Date(don.thoi_gian_het_han).getTime()
    if (Date.now() > hetHanMs) {
      await Ve.update(
        { trang_thai: 'da_huy' },
        { where: { id_don_dat_ve: don.id_don_dat_ve, trang_thai: 'cho_xac_nhan' } }
      )
      await DonDatVe.update(
        { trang_thai: 'het_han' },
        { where: { id_don_dat_ve: don.id_don_dat_ve } }
      )
      don = await BookingRepo.findByMaDatCho(maDatCho)
    }
  }

  return don
}

// Lịch sử đặt vé của tài khoản
const getBookingHistory = (idTaiKhoan) => BookingRepo.findByTaiKhoan(idTaiKhoan)

// Format đơn đặt vé thành dạng trả frontend
const formatDon = (don) => {
  if (!don) return null
  return {
    idDon:         don.id_don_dat_ve,
    maDon:         don.ma_don,
    maDatCho:      don.ma_dat_cho,
    trangThai:     don.trang_thai,
    loaiVe:        don.loai_ve,
    tongTien:      don.tong_tien,
    tienGiam:      don.tien_giam,
    tienThanhToan: don.tien_thanh_toan,
    hoTenLienLac:  don.ho_ten_lien_lac,
    emailDatCho:   don.email_dat_cho,
    sdtDatCho:     don.sdt_dat_cho,
    thoiGianDat:   don.thoi_gian_dat,
    thoiGianHetHan:don.thoi_gian_het_han,
    ve: (don.Ves || []).map(ve => ({
      idVe:          ve.id_ve,
      idChuyen:      ve.id_chuyen,
      soToa:         ve.so_toa_thu_tu,
      soGhe:         ve.so_ghe_trong_toa,
      giaVe:         ve.gia_ve,
      qrVe:          ve.qr_ve || null,
      loaiHanhKhach: ve.loai_hanh_khach,
      trangThai:     ve.trang_thai,
      hanhKhach:     ve.HanhKhach ? {
        hoTen:    ve.HanhKhach.ho_ten,
        ngaySinh: ve.HanhKhach.ngay_sinh,
        cccd:     ve.HanhKhach.cccd,
      } : null,
      chuyen: ve.ChuyenTau ? {
        ngayChay:    ve.ChuyenTau.ngay_chay,
        gioKhoiHanh: ve.ChuyenTau.LichChay?.gio_khoi_hanh,
        gioDen:      ve.ChuyenTau.LichChay?.gio_du_kien_den,
        maTau:       ve.ChuyenTau.LichChay?.Tau?.so_hieu,
        // Ưu tiên GaLen/GaXuong (ga thực hành khách lên/xuống);
        // fallback sang GaDi/GaDen của LichChay (đơn cũ thiếu id_ga_len/id_ga_xuong)
        gaDi:  ve.GaLen?.ten_ga   || ve.ChuyenTau.LichChay?.GaDi?.ten_ga  || '',
        gaDen: ve.GaXuong?.ten_ga || ve.ChuyenTau.LichChay?.GaDen?.ten_ga || '',
      } : null,
    })),
    thanhToan: (don.ThanhToans || []).map(tt => ({
      idThanhToan:   tt.id_thanh_toan,
      maGiaoDich:    tt.ma_giao_dich,
      phuongThuc:    tt.phuong_thuc,
      soTien:        tt.so_tien,
      trangThai:     tt.trang_thai,
      thoiGianTao:   tt.thoi_gian_tao,
      thoiGianTT:    tt.thoi_gian_thanh_toan,
    })),
  }
}

module.exports = { holdSeatsForCheckout, createBooking, lookupBooking, getBookingHistory, formatDon }

File: java-backend/src/main/java/com/klntrain/service/BookingService.java
package com.klntrain.service;

import com.klntrain.dto.request.CreateBookingRequest;
import com.klntrain.entity.*;
import com.klntrain.repository.*;
import com.klntrain.util.AppException;
import com.klntrain.util.CodeGenerator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
public class BookingService {

    private final DonDatVeRepository donDatVeRepo;
    private final VeRepository veRepo;
    private final HanhKhachRepository hanhKhachRepo;
    private final TamGiuGheRepository tamGiuGheRepo;
    private final KhuyenMaiRepository khuyenMaiRepo;

    // Giữ ghế tạm (trước khi điền form)
    @Transactional
    public Map<String, Object> holdSeatsForCheckout(List<CreateBookingRequest.TripInfo> trips) {
        String sessionId = UUID.randomUUID().toString().replace("-", "").substring(0, 16).toUpperCase();
        LocalDateTime hetHan = LocalDateTime.now().plusMinutes(15);

        for (var trip : trips) {
            checkSeatsAvailable(trip.getIdChuyen(), trip.getPassengerSeats(), null);
            for (var ps : trip.getPassengerSeats()) {
                TamGiuGhe tgg = TamGiuGhe.builder()
                    .idChuyen(trip.getIdChuyen())
                    .soToaThuTu(ps.getSoToaThuTu())
                    .soGheTrongToa(ps.getSeatNumber())
                    .sessionId(sessionId)
                    .idGaLen(trip.getIdGaLen())
                    .idGaXuong(trip.getIdGaXuong())
                    .trangThai("dang_giu")
                    .thoiGianHetHan(hetHan)
                    .build();
                tamGiuGheRepo.save(tgg);
            }
        }

        return Map.of("sessionId", sessionId, "hetHan", hetHan);
    }

    @Transactional
    public Map<String, Object> createBooking(CreateBookingRequest req, Long idTaiKhoan) {
        // 1. Validate khuyến mãi
        KhuyenMai km = null;
        BigDecimal tienGiam = BigDecimal.ZERO;
        if (req.getMaKhuyenMai() != null && !req.getMaKhuyenMai().isBlank()) {
            km = khuyenMaiRepo.findByMaKhuyenMai(req.getMaKhuyenMai().toUpperCase())
                .orElseThrow(() -> AppException.badRequest("Mã khuyến mãi không hợp lệ"));
            LocalDate today = LocalDate.now();
            if (today.isBefore(km.getNgayBatDau()) || today.isAfter(km.getNgayHetHan()))
                throw AppException.badRequest("Mã khuyến mãi đã hết hạn");
            if (km.getSoLuong() != null && km.getDaDung() >= km.getSoLuong())
                throw AppException.badRequest("Mã khuyến mãi đã hết lượt");
        }

        // 2. Tính tổng tiền
        BigDecimal tongTienVe = req.getTrips().stream()
            .flatMap(t -> t.getPassengerSeats().stream())
            .map(CreateBookingRequest.PassengerSeat::getSeatPrice)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        BigDecimal phiDichVu = req.getTrips().size() > 1 ? new BigDecimal("40000") : new BigDecimal("20000");
        BigDecimal tongTruocGiam = tongTienVe.add(phiDichVu);

        if (km != null && tongTruocGiam.compareTo(km.getGiaTri_donToiThieu()) >= 0) {
            if ("phan_tram".equals(km.getLoaiGiam())) {
                tienGiam = tongTruocGiam.multiply(km.getGiaTri()).divide(new BigDecimal("100"));
                if (km.getGiamToiDa() != null) tienGiam = tienGiam.min(km.getGiamToiDa());
            } else {
                tienGiam = km.getGiaTri().min(tongTruocGiam);
            }
        }
        BigDecimal tongThanhToan = tongTruocGiam.subtract(tienGiam);

        // 3. Kiểm tra ghế còn trống
        for (var trip : req.getTrips()) {
            checkSeatsAvailable(trip.getIdChuyen(), trip.getPassengerSeats(), trip.getSessionId());
        }

        // 4. Sinh mã
        String maDatCho;
        do { maDatCho = CodeGenerator.genBookingCode(); }
        while (donDatVeRepo.existsByMaDatCho(maDatCho));
        String maDon = CodeGenerator.genOrderCode();

        // 5. Tạo DonDatVe
        var contact = req.getContactInfo();
        var passengers = req.getPassengers();
        DonDatVe don = DonDatVe.builder()
            .maDon(maDon)
            .maDatCho(maDatCho)
            .idTaiKhoan(idTaiKhoan)
            .hoTenLienLac(contact.getHoTen() != null ? contact.getHoTen() : (passengers.isEmpty() ? "" : passengers.get(0).getHoTen()))
            .emailDatCho(contact.getEmail())
            .sdtDatCho(contact.getPhone().replaceAll("\\s", ""))
            .cccd(passengers.isEmpty() ? "000000000" : Optional.ofNullable(passengers.get(0).getCccd()).orElse("000000000"))
            .loaiVe(req.getTrips().size() > 1 ? "khu_hoi" : "mot_chieu")
            .tongTien(tongTruocGiam)
            .tienGiam(tienGiam)
            .tienThanhToan(tongThanhToan)
            .idKhuyenMai(km != null ? km.getIdKhuyenMai() : null)
            .trangThai("cho_thanh_toan")
            .thoiGianHetHan(LocalDateTime.now().plusMinutes(15))
            .build();
        don = donDatVeRepo.save(don);

        // 6. Tạo Ve + HanhKhach
        int pIdx = 0;
        for (var trip : req.getTrips()) {
            for (var ps : trip.getPassengerSeats()) {
                var p = pIdx < passengers.size() ? passengers.get(pIdx++) : passengers.get(passengers.size() - 1);
                String loaiHK = p.isChild() ? "tre_em" : p.isElderly() ? "nguoi_cao_tuoi" : p.isStudent() ? "sinh_vien" : "nguoi_lon";

                LocalDate ngaySinhParsed = null;
                if (p.getNgaySinh() != null && !p.getNgaySinh().isBlank()) {
                    try { ngaySinhParsed = LocalDate.parse(p.getNgaySinh()); } catch (Exception ignored) {}
                }

                LocalDate finalNgaySinh = ngaySinhParsed;
                HanhKhach hk = (ngaySinhParsed != null)
                    ? hanhKhachRepo.findByHoTenAndNgaySinh(p.getHoTen(), ngaySinhParsed).orElse(null)
                    : null;

                if (hk == null) {
                    hk = hanhKhachRepo.save(HanhKhach.builder()
                        .idTaiKhoan(idTaiKhoan)
                        .hoTen(p.getHoTen())
                        .ngaySinh(finalNgaySinh)
                        .cccd(p.getCccd())
                        .loaiHanhKhach(loaiHK)
                        .soDienThoai(contact.getPhone().replaceAll("\\s", ""))
                        .laChinh(pIdx == 1)
                        .build());
                }

                Ve ve = Ve.builder()
                    .donDatVe(don)
                    .hanhKhach(hk)
                    .chuyenTau(new ChuyenTau().toBuilder().idChuyen(trip.getIdChuyen()).build())
                    .soToaThuTu(ps.getSoToaThuTu())
                    .soGheTrongToa(ps.getSeatNumber())
                    .gaLen(trip.getIdGaLen() != null ? GaTau.builder().idGa(trip.getIdGaLen()).build() : null)
                    .gaXuong(trip.getIdGaXuong() != null ? GaTau.builder().idGa(trip.getIdGaXuong()).build() : null)
                    .loaiHanhKhach(loaiHK)
                    .giaVe(ps.getSeatPrice())
                    .trangThai("cho_xac_nhan")
                    .ngayXuatVe(LocalDateTime.now())
                    .build();
                veRepo.save(ve);
            }

            // Cập nhật / tạo TamGiuGhe
            if (trip.getSessionId() != null) {
                tamGiuGheRepo.releaseBySessionId(trip.getSessionId());
            }
        }

        // 7. Tăng lượt dùng khuyến mãi
        if (km != null) {
            km.setDaDung(km.getDaDung() == null ? 1 : km.getDaDung() + 1);
            khuyenMaiRepo.save(km);
        }

        return Map.of(
            "maDon", maDon,
            "maDatCho", maDatCho,
            "idDon", don.getIdDonDatVe(),
            "tongThanhToan", tongThanhToan,
            "tienGiam", tienGiam
        );
    }

    public Map<String, Object> lookupBooking(String maDatCho, String email, String phone) {
        DonDatVe don = donDatVeRepo.findByMaDatCho(maDatCho)
            .orElse(null);
        if (don == null) return null;

        boolean emailOk = don.getEmailDatCho().equalsIgnoreCase(email.strip());
        String phoneClean = don.getSdtDatCho().replaceAll("\\D", "");
        String inputPhone = phone.strip().replaceAll("\\D", "");
        boolean phoneOk = phoneClean.equals(inputPhone) ||
            phoneClean.substring(Math.max(0, phoneClean.length() - 9)).equals(inputPhone.substring(Math.max(0, inputPhone.length() - 9)));

        if (!emailOk || !phoneOk) return null;

        return formatDon(don);
    }

    public List<Map<String, Object>> getBookingHistory(Long idTaiKhoan) {
        return donDatVeRepo.findByIdTaiKhoanOrderByThoiGianDatDesc(idTaiKhoan)
            .stream().map(this::formatDon).toList();
    }

    public Map<String, Object> formatDon(DonDatVe don) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("idDon", don.getIdDonDatVe());
        m.put("maDon", don.getMaDon());
        m.put("maDatCho", don.getMaDatCho());
        m.put("trangThai", don.getTrangThai());
        m.put("loaiVe", don.getLoaiVe());
        m.put("tongTien", don.getTongTien());
        m.put("tienGiam", don.getTienGiam());
        m.put("tienThanhToan", don.getTienThanhToan());
        m.put("hoTenLienLac", don.getHoTenLienLac());
        m.put("emailDatCho", don.getEmailDatCho());
        m.put("sdtDatCho", don.getSdtDatCho());
        m.put("thoiGianDat", don.getThoiGianDat());
        m.put("thoiGianHetHan", don.getThoiGianHetHan());

        List<Map<String, Object>> veData = new ArrayList<>();
        if (don.getVeList() != null) {
            for (Ve ve : don.getVeList()) {
                Map<String, Object> vm = new LinkedHashMap<>();
                vm.put("idVe", ve.getIdVe());
                vm.put("idChuyen", ve.getChuyenTau() != null ? ve.getChuyenTau().getIdChuyen() : null);
                vm.put("soToa", ve.getSoToaThuTu());
                vm.put("soGhe", ve.getSoGheTrongToa());
                vm.put("giaVe", ve.getGiaVe());
                vm.put("loaiHanhKhach", ve.getLoaiHanhKhach());
                vm.put("trangThai", ve.getTrangThai());
                if (ve.getHanhKhach() != null) {
                    vm.put("hanhKhach", Map.of(
                        "hoTen", ve.getHanhKhach().getHoTen(),
                        "ngaySinh", ve.getHanhKhach().getNgaySinh() != null ? ve.getHanhKhach().getNgaySinh().toString() : "",
                        "cccd", ve.getHanhKhach().getCccd() != null ? ve.getHanhKhach().getCccd() : ""
                    ));
                }
                veData.add(vm);
            }
        }
        m.put("ve", veData);
        return m;
    }

    private void checkSeatsAvailable(Long idChuyen, List<CreateBookingRequest.PassengerSeat> pSeats, String sessionId) {
        for (var ps : pSeats) {
            long cnt = veRepo.countActiveByChuyenAndGhe(idChuyen, ps.getSoToaThuTu(), ps.getSeatNumber());
            if (cnt > 0) throw AppException.conflict("Ghế " + ps.getSeatNumber() + " toa " + ps.getSoToaThuTu() + " đã được đặt");

            // Kiểm tra TamGiuGhe
            List<TamGiuGhe> holds = tamGiuGheRepo.findConflictingHolds(
                idChuyen, ps.getSoToaThuTu(), List.of(ps.getSeatNumber()),
                LocalDateTime.now(), sessionId
            );
            if (!holds.isEmpty()) throw AppException.conflict("Ghế đang được giữ bởi người khác");
        }
    }
}
